packages/firestore/src/core/sync_engine.ts

```diff
@@ -381,92 +363,81 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
    *
    * The promise returned is resolved when the transaction is fully committed.
    */
-  runTransaction<T>(
+  async runTransaction<T>(
     updateFunction: (transaction: Transaction) => Promise<T>,
     retries = 5
   ): Promise<T> {
     assert(retries >= 0, 'Got negative number of retries for transaction.');
     const transaction = this.remoteStore.createTransaction();
-    const wrappedUpdateFunction = (): Promise<T> => {
-      try {
-        const userPromise = updateFunction(transaction);
-        if (
-          isNullOrUndefined(userPromise) ||
-          !userPromise.catch ||
-          !userPromise.then
-        ) {
-          return Promise.reject<T>(
-            Error('Transaction callback must return a Promise')
-          );
-        }
-        return userPromise.catch(e => {
-          return Promise.reject<T>(this.wrapUpdateFunctionError(e));
-        });
-      } catch (e) {
-        return Promise.reject<T>(this.wrapUpdateFunctionError(e));
+    let result: T;
+    try {
+      const userPromise = updateFunction(transaction);
+      if (
+        isNullOrUndefined(userPromise) ||
+        !userPromise.catch ||
+        !userPromise.then
+      ) {
+        return Promise.reject<T>(
+          Error('Transaction callback must return a Promise')
+        );
       }
-    };
-    return wrappedUpdateFunction().then(result => {
-      return transaction
-        .commit()
-        .then(() => {
-          return result;
-        })
-        .catch(error => {
-          if (retries === 0) {
-            return Promise.reject<T>(error);
-          }
-          // TODO(klimt): Put in a retry delay?
-          return this.runTransaction(updateFunction, retries - 1);
-        });
-    });
+      userPromise.catch(error => {
```

Comment:
This dance seems unnecessary. `userPromise` is already defined as returning `Promise<T>`, so if it fails the result will already be of that type. Meanwhile, the `return Promise.reject<T>(error);` is local only to the function you passed to catch, so you've effectively passed an identity function to catch.

Maybe what's going on is that previously, this was calling `wrapUpdateFunctionError(error)` so it did serve some purpose. Now that you've removed that translation step this can entirely go away.